constraint/pkg/apis/templates/transform.go

```diff
@@ -0,0 +1,84 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	"fmt"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+var trueBool = true
+
+// AddPreserveUnknownFields recurses through an *apiextensionsv1beta1.JSONSchemaProps
+// data structure, adding `x-kubernetes-preserve-unknown-fields: true` at every level
+// that type is equal to "object", "array", or is undefined.
+func AddPreserveUnknownFields(sch *apiextensionsv1beta1.JSONSchemaProps) error {
+	// An object can have values not described in the schema.  A blank Type could be
+	// anything, including an object, so we should treat it the same as those types.
+	// typeWithFields := sch.Type == "" || sch.Type == "object"
+	typeWithFields := sch.Type == "" || sch.Type == "object"
+
+	// Don't override an already set XPreserveUnknownFields
+	if typeWithFields && sch.XPreserveUnknownFields == nil {
+		sch.XPreserveUnknownFields = &trueBool
+	}
+
+	if sch.Properties != nil {
+		for k, v := range sch.Properties {
```

Comment:
nit: to avoid the setting back onto original structure, use:

`AddPreserveUnknownFields(&sch.Properties[k])`

Reply:
That doesn't seem to compile:

```
invalid operation: cannot take address of sch.Properties[k] (map index expression of type v1beta1.JSONSchemaProps)
```

[Stackoverflow says](https://stackoverflow.com/a/32496031) this is actually a design feature of golang that's made to prevent the creation of a dangling pointer.

Reply:
I guess there's still some things in go that will surprise me.  Perhaps use a temp variable for the pointer instead?

```
for k := range ... {
  v := &sch.Properties
}
```

Reply:
This yields the same error:

```
		for k := range sch.Properties {
			v := &sch.Properties[k] // invalid operation: cannot take address of sch.Properties[k]...
			if err := AddPreserveUnknownFields(v); err != nil {
				return err
			}
		}
```